File: src/utils/quantization_mapper.py
"""
Quantization configuration mapper for converting four-field quant config
to runtime-specific CLI arguments (vLLM, SGLang, TensorRT-LLM).
"""

import logging
from typing import Dict, Any, Optional

logger = logging.getLogger(__name__)


# Quantization presets (optional, for convenience)
QUANTIZATION_PRESETS = {
    "default": {
        "gemm_dtype": "auto",
        "kvcache_dtype": "auto",
        "attention_dtype": "auto",
        "moe_dtype": "auto"
    },
    "kv-cache-fp8": {
        "gemm_dtype": "auto",
        "kvcache_dtype": "fp8_e5m2",
        "attention_dtype": "auto",
        "moe_dtype": "auto"
    },
    "dynamic-fp8": {
        "gemm_dtype": "fp8",
        "kvcache_dtype": "fp8_e5m2",
        "attention_dtype": "fp8",
        "moe_dtype": "fp8"
    },
    "bf16-stable": {
        "gemm_dtype": "bfloat16",
        "kvcache_dtype": "fp8_e5m2",
        "attention_dtype": "auto",
        "moe_dtype": "auto"
    },
    "aggressive-moe": {
        "gemm_dtype": "bfloat16",
        "kvcache_dtype": "fp8_e5m2",
        "attention_dtype": "fp8",
        "moe_dtype": "w4afp8"
    }
}


def expand_preset(preset_name: str) -> Dict[str, str]:
    """Expand preset name to full quantization configuration."""
    if preset_name not in QUANTIZATION_PRESETS:
        raise ValueError(f"Unknown quantization preset: {preset_name}. "
                         f"Available presets: {list(QUANTIZATION_PRESETS.keys())}")
    return QUANTIZATION_PRESETS[preset_name].copy()


def validate_quant_config(config: Dict[str, Any]) -> tuple[bool, str]:
    """
    Validate quantization configuration.

    Returns:
        (is_valid, error_message)
    """
    if not config:
        return True, "Valid (empty config)"

    # Check if it's a preset reference
    if "preset" in config:
        preset_name = config["preset"]
        if preset_name not in QUANTIZATION_PRESETS:
            return False, f"Unknown preset: {preset_name}"
        return True, "Valid preset"

    # Validate individual fields
    valid_gemm_dtypes = ["auto", "float16", "bfloat16", "float32", "fp8", "int8"]
    valid_kvcache_dtypes = ["auto", "fp16", "bfloat16", "fp8", "fp8_e5m2", "fp8_e4m3", "int8", "int4"]
    valid_attention_dtypes = ["auto", "float16", "bfloat16", "fp8", "fp8_e5m2", "fp8_e4m3", "fp8_block"]
    valid_moe_dtypes = ["auto", "float16", "bfloat16", "fp8", "w4afp8", "mxfp4", "int8"]

    gemm_dtype = config.get("gemm_dtype", "auto")
    if gemm_dtype not in valid_gemm_dtypes:
        return False, f"Invalid gemm_dtype: {gemm_dtype}. Must be one of {valid_gemm_dtypes}"

    kvcache_dtype = config.get("kvcache_dtype", "auto")
    if kvcache_dtype not in valid_kvcache_dtypes:
        return False, f"Invalid kvcache_dtype: {kvcache_dtype}. Must be one of {valid_kvcache_dtypes}"

    attention_dtype = config.get("attention_dtype", "auto")
    if attention_dtype not in valid_attention_dtypes:
        return False, f"Invalid attention_dtype: {attention_dtype}. Must be one of {valid_attention_dtypes}"

    moe_dtype = config.get("moe_dtype", "auto")
    if moe_dtype not in valid_moe_dtypes:
        return False, f"Invalid moe_dtype: {moe_dtype}. Must be one of {valid_moe_dtypes}"

    return True, "Valid"


def resolve_quant_config(config: Optional[Dict[str, Any]]) -> Dict[str, str]:
    """
    Resolve quantization configuration, expanding presets if needed.

    Args:
        config: Quantization config (can contain 'preset' or explicit fields)

    Returns:
        Resolved config with explicit gemm_dtype, kvcache_dtype, attention_dtype, moe_dtype
    """
    if not config:
        # Default: no quantization
        return {
            "gemm_dtype": "auto",
            "kvcache_dtype": "auto",
            "attention_dtype": "auto",
            "moe_dtype": "auto"
        }

    # If preset is specified, expand it
    if "preset" in config:
        base_config = expand_preset(config["preset"])
        # Allow overriding preset values with explicit fields
        base_config.update({k: v for k, v in config.items() if k != "preset"})
        return base_config

    # Use explicit fields with defaults
    return {
        "gemm_dtype": config.get("gemm_dtype", "auto"),
        "kvcache_dtype": config.get("kvcache_dtype", "auto"),
        "attention_dtype": config.get("attention_dtype", "auto"),
        "moe_dtype": config.get("moe_dtype", "auto")
    }


def should_apply_dynamic_fp8(
    gemm_dtype: str,
    model_quantization: Optional[str] = None
) -> bool:
    """
    Determine if dynamic FP8 (W8A8) quantization should be applied.

    Returns False if model is already quantized with offline methods.
    """
    if gemm_dtype != "fp8":
        return False

    # List of offline quantization methods
    offline_quant_methods = [
        "awq", "gptq", "gguf", "squeezellm", "marlin",
        "nvfp4", "fp8", "bitsandbytes", "hqq"
    ]

    if model_quantization and model_quantization in offline_quant_methods:
        logger.warning(
            f"Ignoring gemm_dtype='fp8': model already quantized with {model_quantization}. "
            f"Dynamic FP8 only applies to unquantized FP16/BF16 models."
        )
        return False

    return True


def map_to_vllm_args(
    quant_config: Dict[str, str],
    model_quantization: Optional[str] = None
) -> Dict[str, str]:
    """
    Map four-field quantization config to vLLM CLI arguments.

    Args:
        quant_config: Resolved quant config with gemm_dtype, kvcache_dtype, attention_dtype, moe_dtype
        model_quantization: Detected model quantization method (e.g., "awq", "gptq", None)

    Returns:
        Dictionary of vLLM CLI arguments (e.g., {"--dtype": "auto", "--kv-cache-dtype": "fp8"})
    """
    args = {}

    gemm_dtype = quant_config["gemm_dtype"]
    kvcache_dtype = quant_config["kvcache_dtype"]
    attention_dtype = quant_config["attention_dtype"]
    # moe_dtype not supported by vLLM (uses gemm_dtype)

    # GEMM dtype
    if should_apply_dynamic_fp8(gemm_dtype, model_quantization):
        # Dynamic FP8 quantization (W8A8)
        args["--quantization"] = "fp8"
        args["--dtype"] = "auto"
    elif gemm_dtype == "int8":
        args["--quantization"] = "int8"
        args["--dtype"] = "auto"
    elif gemm_dtype != "auto":
        # Explicit dtype (float16, bfloat16, float32)
        args["--dtype"] = gemm_dtype
    else:
        # Auto: let vLLM decide
        args["--dtype"] = "auto"

    # KV cache dtype
    if kvcache_dtype != "auto":
        args["--kv-cache-dtype"] = kvcache_dtype

    # Attention dtype
    # vLLM doesn't support explicit attention dtype control (uses GEMM dtype)
    if attention_dtype not in ["auto", gemm_dtype]:
        logger.warning(
            f"vLLM does not support separate attention_dtype. "
            f"Requested '{attention_dtype}' will fall back to gemm_dtype '{gemm_dtype}'."
        )

    return args


def map_to_sglang_args(
    quant_config: Dict[str, str],
    model_quantization: Optional[str] = None
) -> Dict[str, str]:
    """
    Map four-field quantization config to SGLang CLI arguments.

    Args:
        quant_config: Resolved quant config with gemm_dtype, kvcache_dtype, attention_dtype, moe_dtype
        model_quantization: Detected model quantization method (e.g., "awq", "gptq", None)

    Returns:
        Dictionary of SGLang CLI arguments
    """
    args = {}

    gemm_dtype = quant_config["gemm_dtype"]
    kvcache_dtype = quant_config["kvcache_dtype"]
    attention_dtype = quant_config["attention_dtype"]
    moe_dtype = quant_config["moe_dtype"]

    # GEMM dtype
    # MoE dtype can override GEMM quantization for MoE models
    if moe_dtype in ["w4afp8", "mxfp4"]:
        # MoE-specific quantization takes precedence
        args["--quantization"] = moe_dtype
        args["--dtype"] = "auto"

        # Set MoE backend
        if moe_dtype == "w4afp8":
            args["--moe-runner-backend"] = "flashinfer_mxfp4"
        elif moe_dtype == "mxfp4":
            args["--moe-runner-backend"] = "flashinfer_mxfp4"
    elif should_apply_dynamic_fp8(gemm_dtype, model_quantization):
        # Dynamic FP8 quantization (W8A8)
        args["--quantization"] = "fp8"
        args["--dtype"] = "auto"
    elif gemm_dtype == "int8":
        args["--quantization"] = "int8"
        args["--dtype"] = "auto"
    elif gemm_dtype != "auto":
        # Explicit dtype (float16, bfloat16)
        args["--dtype"] = gemm_dtype
    else:
        # Auto: let SGLang decide
        args["--dtype"] = "auto"

    # KV cache dtype
    if kvcache_dtype != "auto":
        args["--kv-cache-dtype"] = kvcache_dtype

    # Attention dtype
    if attention_dtype in ["fp8", "fp8_e5m2", "fp8_e4m3"]:
        # Enable FP8 attention with FlashInfer backend
        args["--attention-backend"] = "flashinfer"
        # FP8 attention is enabled automatically when using FlashInfer + FP8 KV cache

    # MoE dtype (if FP8 but not w4afp8/mxfp4)
    if moe_dtype == "fp8" and "--quantization" not in args:
        # Enable FP8 MoE with FlashInfer backend
        args["--moe-runner-backend"] = "flashinfer_cutlass"

    return args


def map_to_tensorrt_llm_args(
    quant_config: Dict[str, str],
    model_quantization: Optional[str] = None
) -> Dict[str, str]:
    """
    Map four-field quantization config to TensorRT-LLM arguments.

    Args:
        quant_config: Resolved quant config with gemm_dtype, kvcache_dtype, attention_dtype, moe_dtype
        model_quantization: Detected model quantization method

    Returns:
        Dictionary of TensorRT-LLM arguments
    """
    args = {}

    gemm_dtype = quant_config["gemm_dtype"]
    kvcache_dtype = quant_config["kvcache_dtype"]
    attention_dtype = quant_config["attention_dtype"]
    # moe_dtype not supported by TensorRT-LLM (uses GEMM dtype)

    # GEMM dtype
    if should_apply_dynamic_fp8(gemm_dtype, model_quantization):
        args["--quant-algo"] = "FP8"
    elif gemm_dtype == "int8":
        args["--quant-algo"] = "INT8"
    # Note: TensorRT-LLM auto-detects model dtype, no explicit --dtype flag

    # KV cache dtype
    if "fp8" in kvcache_dtype:
        args["--kv-cache-quant-algo"] = "FP8"
    elif kvcache_dtype == "int8":
        args["--kv-cache-quant-algo"] = "INT8"
    elif kvcache_dtype == "int4":
        args["--kv-cache-quant-algo"] = "INT4"

    # Attention dtype (FMHA quantization)
    if attention_dtype in ["fp8", "fp8_e5m2", "fp8_e4m3"]:
        args["--fmha-quant-algo"] = "FP8"
    elif attention_dtype == "fp8_block":
        args["--fmha-quant-algo"] = "FP8_BLOCK"

    return args


def merge_parameters(
    quant_args: Dict[str, str],
    user_parameters: Dict[str, Any]
) -> Dict[str, str]:
    """
    Merge quantization-derived arguments with user-specified parameters.
    User parameters have higher priority and will override quant_config-derived values.

    Args:
        quant_args: Arguments derived from quant_config
        user_parameters: User-specified parameters from Task.parameters

    Returns:
        Merged parameter dictionary with user parameters taking precedence
    """
    # Start with quant_config-derived args
    merged = quant_args.copy()

    # Override with user parameters
    # Convert user parameters to CLI format if needed
    for key, value in user_parameters.items():
        # Handle both formats: "dtype" and "--dtype"
        cli_key = key if key.startswith("--") else f"--{key}"

        # If user explicitly set this parameter, override quant_config
        if isinstance(value, list):
            # If it's a list (parameter grid), use the first value for now
            # The orchestrator will handle grid expansion
            if value:
                merged[cli_key] = str(value[0])
        else:
            merged[cli_key] = str(value)

    return merged


def get_runtime_args(
    runtime: str,
    quant_config: Optional[Dict[str, Any]],
    user_parameters: Optional[Dict[str, Any]] = None,
    model_quantization: Optional[str] = None
) -> Dict[str, str]:
    """
    Get runtime-specific CLI arguments from quantization config and user parameters.

    Args:
        runtime: Runtime engine ("vllm", "sglang", "tensorrt_llm")
        quant_config: Quantization configuration (can be None, preset, or explicit fields)
        user_parameters: User-specified parameters (optional, takes priority over quant_config)
        model_quantization: Detected offline model quantization (e.g., "awq", "gptq")

    Returns:
        Dictionary of CLI arguments ready to be passed to the runtime

    Example:
        >>> get_runtime_args(
        ...     "sglang",
        ...     {"gemm_dtype": "fp8", "kvcache_dtype": "fp8_e5m2"},
        ...     {"tp-size": [1, 2], "mem-fraction-static": 0.85}
        ... )
        {"--quantization": "fp8", "--dtype": "auto", "--kv-cache-dtype": "fp8_e5m2",
         "--tp-size": "1", "--mem-fraction-static": "0.85"}
    """
    # Validate and resolve quant config
    is_valid, msg = validate_quant_config(quant_config or {})
    if not is_valid:
        raise ValueError(f"Invalid quantization config: {msg}")

    resolved_config = resolve_quant_config(quant_config)

    # Map to runtime-specific arguments
    runtime_lower = runtime.lower()
    if runtime_lower == "vllm":
        quant_args = map_to_vllm_args(resolved_config, model_quantization)
    elif runtime_lower == "sglang":
        quant_args = map_to_sglang_args(resolved_config, model_quantization)
    elif runtime_lower in ["tensorrt_llm", "trtllm", "tensorrt-llm"]:
        quant_args = map_to_tensorrt_llm_args(resolved_config, model_quantization)
    else:
        logger.warning(f"Unknown runtime: {runtime}. Returning empty quant args.")
        quant_args = {}

    # Merge with user parameters (user parameters take priority)
    if user_parameters:
        return merge_parameters(quant_args, user_parameters)
    else:
        return quant_args

File: frontend/src/utils/quantizationMapper.ts
// TypeScript port of quantization_mapper.py
// Maps quantization configuration to runtime-specific CLI arguments

import { QuantizationConfig } from '../types/api';

// Resolved config with only string types (arrays resolved to first value)
interface ResolvedQuantConfig {
	gemm_dtype: string;
	kvcache_dtype: string;
	attention_dtype: string;
	moe_dtype: string;
}

// Quantization presets matching backend implementation
export const QUANTIZATION_PRESETS: Record<string, Omit<QuantizationConfig, 'preset' | 'presets'>> = {
	'default': {
		gemm_dtype: 'auto',
		kvcache_dtype: 'auto',
		attention_dtype: 'auto',
		moe_dtype: 'auto'
	},
	'kv-cache-fp8': {
		gemm_dtype: 'auto',
		kvcache_dtype: 'fp8_e5m2',
		attention_dtype: 'auto',
		moe_dtype: 'auto'
	},
	'dynamic-fp8': {
		gemm_dtype: 'fp8',
		kvcache_dtype: 'fp8_e5m2',
		attention_dtype: 'fp8',
		moe_dtype: 'auto'
	},
	'bf16-stable': {
		gemm_dtype: 'bfloat16',
		kvcache_dtype: 'fp8_e5m2',
		attention_dtype: 'auto',
		moe_dtype: 'auto'
	},
	'aggressive-moe': {
		gemm_dtype: 'auto',
		kvcache_dtype: 'fp8_e5m2',
		attention_dtype: 'fp8',
		moe_dtype: 'w4afp8'
	}
};

// Expand preset to full config
export function expandPreset(presetName: string): ResolvedQuantConfig {
	const preset = QUANTIZATION_PRESETS[presetName];
	if (!preset) {
		return {
			gemm_dtype: 'auto',
			kvcache_dtype: 'auto',
			attention_dtype: 'auto',
			moe_dtype: 'auto'
		};
	}

	// Convert string | string[] to string
	const resolveValue = (val: string | string[] | undefined): string => {
		if (!val) return 'auto';
		if (Array.isArray(val)) return val[0] || 'auto';
		return val;
	};

	return {
		gemm_dtype: resolveValue(preset.gemm_dtype),
		kvcache_dtype: resolveValue(preset.kvcache_dtype),
		attention_dtype: resolveValue(preset.attention_dtype),
		moe_dtype: resolveValue(preset.moe_dtype)
	};
}

// Resolve config to flat structure (takes first value if array)
export function resolveQuantConfig(config: QuantizationConfig): ResolvedQuantConfig {
	// Multi-preset mode - use first preset for display
	if (config.presets && config.presets.length > 0) {
		return expandPreset(config.presets[0]);
	}

	// Single preset mode
	if (config.preset) {
		return expandPreset(config.preset);
	}

	// Custom mode or empty - take first value if array
	const resolveField = (value: string | string[] | undefined): string => {
		if (!value) return 'auto';
		if (Array.isArray(value)) return value[0] || 'auto';
		return value;
	};

	return {
		gemm_dtype: resolveField(config.gemm_dtype),
		kvcache_dtype: resolveField(config.kvcache_dtype),
		attention_dtype: resolveField(config.attention_dtype),
		moe_dtype: resolveField(config.moe_dtype)
	};
}

// Map to vLLM arguments
export function mapToVllmArgs(config: ResolvedQuantConfig): Record<string, string> {
	const args: Record<string, string> = {};

	// GEMM dtype -> quantization method
	if (config.gemm_dtype === 'fp8') {
		args['--quantization'] = 'fp8';
		args['--dtype'] = 'auto';
	} else if (config.gemm_dtype === 'int8') {
		args['--quantization'] = 'int8';
		args['--dtype'] = 'auto';
	} else if (config.gemm_dtype && config.gemm_dtype !== 'auto') {
		args['--dtype'] = config.gemm_dtype;
	}

	// KV cache dtype
	if (config.kvcache_dtype && config.kvcache_dtype !== 'auto') {
		args['--kv-cache-dtype'] = config.kvcache_dtype;
	}

	// Note: vLLM doesn't support separate attention dtype
	// Note: vLLM has limited MoE dtype control

	return args;
}

// Map to SGLang arguments
export function mapToSglangArgs(config: ResolvedQuantConfig): Record<string, string> {
	const args: Record<string, string> = {};

	// MoE dtype can override GEMM quantization
	if (config.moe_dtype === 'w4afp8') {
		args['--quantization'] = 'w4afp8';
		args['--moe-runner-backend'] = 'flashinfer_cutlass';
	} else if (config.moe_dtype === 'mxfp4') {
		args['--quantization'] = 'mxfp4';
		args['--moe-runner-backend'] = 'flashinfer_mxfp4';
	} else {
		// GEMM dtype
		if (config.gemm_dtype === 'fp8') {
			args['--quantization'] = 'fp8';
			args['--dtype'] = 'auto';
		} else if (config.gemm_dtype === 'int8') {
			args['--quantization'] = 'int8';
			args['--dtype'] = 'auto';
		} else if (config.gemm_dtype && config.gemm_dtype !== 'auto') {
			args['--dtype'] = config.gemm_dtype;
		}
	}

	// KV cache dtype
	if (config.kvcache_dtype && config.kvcache_dtype !== 'auto') {
		args['--kv-cache-dtype'] = config.kvcache_dtype;
	}

	// Attention backend for FP8
	if (config.attention_dtype && ['fp8', 'fp8_e5m2', 'fp8_e4m3'].includes(config.attention_dtype)) {
		args['--attention-backend'] = 'flashinfer';
	}

	return args;
}

// Map to TensorRT-LLM arguments
export function mapToTensorrtLlmArgs(config: ResolvedQuantConfig): Record<string, string> {
	const args: Record<string, string> = {};

	// GEMM dtype
	if (config.gemm_dtype === 'fp8') {
		args['--quant-algo'] = 'FP8';
	} else if (config.gemm_dtype === 'int8') {
		args['--quant-algo'] = 'W8A8_SQ_PER_CHANNEL';
	} else if (config.gemm_dtype && config.gemm_dtype !== 'auto') {
		args['--dtype'] = config.gemm_dtype;
	}

	// KV cache dtype
	if (config.kvcache_dtype && config.kvcache_dtype !== 'auto') {
		if (config.kvcache_dtype.includes('fp8')) {
			args['--kv-cache-quant-algo'] = 'FP8';
		} else if (config.kvcache_dtype === 'int8') {
			args['--kv-cache-quant-algo'] = 'INT8';
		} else if (config.kvcache_dtype === 'int4') {
			args['--kv-cache-quant-algo'] = 'INT4';
		}
	}

	// Attention dtype (FMHA quantization)
	if (config.attention_dtype && ['fp8', 'fp8_e5m2', 'fp8_e4m3'].includes(config.attention_dtype)) {
		args['--fmha-quant-algo'] = 'FP8';
	} else if (config.attention_dtype === 'fp8_block') {
		args['--fmha-quant-algo'] = 'FP8_BLOCK';
	}

	// Note: TensorRT-LLM doesn't support separate MoE dtype

	return args;
}

// Get runtime arguments for specific engine
export function getRuntimeArgs(
	runtime: string,
	config: QuantizationConfig
): Record<string, string> {
	const resolvedConfig = resolveQuantConfig(config);

	const runtimeLower = runtime.toLowerCase();

	if (runtimeLower === 'vllm') {
		return mapToVllmArgs(resolvedConfig);
	} else if (runtimeLower === 'sglang') {
		return mapToSglangArgs(resolvedConfig);
	} else if (runtimeLower === 'tensorrt-llm' || runtimeLower === 'tensorrt_llm') {
		return mapToTensorrtLlmArgs(resolvedConfig);
	}

	return {};
}

// Generate all combinations from arrays
function generateCombinations(config: QuantizationConfig): ResolvedQuantConfig[] {
	// Get arrays for each field, defaulting to ['auto'] if empty or undefined
	const getValues = (field: string | string[] | undefined): string[] => {
		if (Array.isArray(field)) {
			return field.length > 0 ? field : ['auto'];
		}
		return [field || 'auto'];
	};

	const gemmValues = getValues(config.gemm_dtype);
	const kvcacheValues = getValues(config.kvcache_dtype);
	const attentionValues = getValues(config.attention_dtype);
	const moeValues = getValues(config.moe_dtype);

	const combinations: ResolvedQuantConfig[] = [];

	// Generate all combinations using nested loops
	for (const gemm of gemmValues) {
		for (const kvcache of kvcacheValues) {
			for (const attention of attentionValues) {
				for (const moe of moeValues) {
					combinations.push({
						gemm_dtype: gemm,
						kvcache_dtype: kvcache,
						attention_dtype: attention,
						moe_dtype: moe
					});
				}
			}
		}
	}

	return combinations;
}

// Get all runtime argument combinations for a config
export function getAllRuntimeArgCombinations(
	runtime: string,
	config: QuantizationConfig,
	maxCombinations: number = 10
): { combinations: Record<string, string>[]; total: number; truncated: boolean } {
	// Handle preset/multi-preset modes
	if (config.presets && config.presets.length > 0) {
		// For multi-preset, each preset is a separate combination
		const combinations = config.presets.map(preset => {
			const resolvedConfig = expandPreset(preset);
			return mapConfigToArgs(runtime, resolvedConfig);
		});

		// Deduplicate combinations
		const uniqueCombinations = deduplicateCombinations(combinations);
		const limitedCombinations = uniqueCombinations.slice(0, maxCombinations);

		return {
			combinations: limitedCombinations,
			total: uniqueCombinations.length,
			truncated: uniqueCombinations.length > maxCombinations
		};
	}

	if (config.preset) {
		// Single preset - just one combination
		const resolvedConfig = expandPreset(config.preset);
		return {
			combinations: [mapConfigToArgs(runtime, resolvedConfig)],
			total: 1,
			truncated: false
		};
	}

	// Custom mode - generate all combinations
	const allCombinations = generateCombinations(config);
	const argCombinations = allCombinations.map(resolved =>
		mapConfigToArgs(runtime, resolved)
	);

	// Deduplicate combinations
	const uniqueCombinations = deduplicateCombinations(argCombinations);
	const limitedCombinations = uniqueCombinations.slice(0, maxCombinations);

	return {
		combinations: limitedCombinations,
		total: uniqueCombinations.length,
		truncated: uniqueCombinations.length > maxCombinations
	};
}

// Deduplicate combinations based on their argument content
function deduplicateCombinations(combinations: Record<string, string>[]): Record<string, string>[] {
	const seen = new Set<string>();
	const unique: Record<string, string>[] = [];

	for (const combo of combinations) {
		// Create a canonical string representation by sorting keys
		const sortedEntries = Object.entries(combo).sort(([keyA], [keyB]) => keyA.localeCompare(keyB));
		const signature = JSON.stringify(sortedEntries);

		if (!seen.has(signature)) {
			seen.add(signature);
			unique.push(combo);
		}
	}

	return unique;
}

// Helper to map a resolved config to args based on runtime
function mapConfigToArgs(runtime: string, config: ResolvedQuantConfig): Record<string, string> {
	const runtimeLower = runtime.toLowerCase();

	if (runtimeLower === 'vllm') {
		return mapToVllmArgs(config);
	} else if (runtimeLower === 'sglang') {
		return mapToSglangArgs(config);
	} else if (runtimeLower === 'tensorrt-llm' || runtimeLower === 'tensorrt_llm') {
		return mapToTensorrtLlmArgs(config);
	}

	return {};
}

// Format arguments for display
export function formatArgsForDisplay(args: Record<string, string>): string {
	if (Object.keys(args).length === 0) {
		return 'No quantization arguments (using defaults)';
	}

	return Object.entries(args)
		.map(([key, value]) => `${key} ${value}`)
		.join(' ');
}

// Format multiple combinations for display
export function formatMultipleCombinationsForDisplay(
	combinations: Record<string, string>[],
	total: number,
	truncated: boolean
): string {
	if (combinations.length === 0) {
		return 'No quantization arguments (using defaults)';
	}

	if (combinations.length === 1) {
		return formatArgsForDisplay(combinations[0]);
	}

	const lines = combinations.map((args, index) => {
		const formattedArgs = formatArgsForDisplay(args);
		return `[${index + 1}] ${formattedArgs}`;
	});

	if (truncated) {
		lines.push(`... and ${total - combinations.length} more combinations`);
	}

	return lines.join('\n');
}
